### PR TITLE
Make python-rhsm Python 3 compatible

### PR DESCRIPTION
**Pull request by kahowell**
This requires changes in #184, so that PR should be reviewed first. Also, I tried to break this up into smaller commits that could be reviewed somewhat independently (essentially C-bindings, setup.py, src, and test), though I'm thinking we'll likely just squash if all changes look good.

### CODE REVIEW

**Comment by barnabycourt** [comment]
@kahowell please squash this PR and resolve the merge conflicts

**Comment by kahowell** [comment]
@barnabycourt, done.

**Comment by kahowell** [comment]
retest this please

File: src/rhsm/certificate.py [hunk 1]
```diff
@@ -562,7 +562,7 @@ def write(self, pem_path):
         f.write(self.content)
         self.path = pem_path
         f.close()
-        os.chmod(pem_path, 0600)
+        os.chmod(pem_path, 0o600)
```
Comment (barnabycourt) [inline src/rhsm/certificate.py]: Is the 0o on purpose? Also, could the constants be used in place? https://docs.python.org/3/library/os.html#os.chmod

Comment (kahowell) [inline src/rhsm/certificate.py]: Yes. This change was done by 2to3. See https://www.python.org/dev/peps/pep-3127/#removal-of-old-octal-syntax

We *could* use the constants, but personally, I find them far less readable than the octal.

File: setup.py [hunk 2]
```diff
@@ -44,7 +44,7 @@ def initialize_options(self):
     def get_git_describe(self):
         cmd = ["git", "describe"]
         process = Popen(cmd, stdout=PIPE)
-        output = process.communicate()[0].strip()
+        output = process.communicate()[0].strip().decode('utf-8')
```
Comment (kahowell) [inline setup.py]: It returns bytes (see https://docs.python.org/3/library/subprocess.html#subprocess.Popen.communicate). We *could* set `universal_newlines=True` in the Popen command instead but, meh.

File: src/rhsm/connection.py [hunk 3]
```diff
@@ -533,7 +510,7 @@ def _request(self, request_type, method, info=None):
         if not len(result['content']):
             return None
 
-        return json.loads(result['content'], object_hook=self._decode_dict)
+        return json.loads(result['content'])
```
Comment (awood) [inline src/rhsm/connection.py]: Do we not still need `_decode_dict` and `_decode_list` in Python 2 to get everything converted to UTF8 or is that being handled elsewhere?

**Comment by awood** [comment]
Overall looks good, but I would like to get a Jenkins Python 3.0 build created and added to the default checks as part of this (enable it by default immediately after this hits master).

File: src/rhsm/connection.py (hunk 3 above)
Comment (kahowell) [inline src/rhsm/connection.py]: No, we don't need it anywhere except for after the read call (I just pushed a quick change; I missed one of the read calls, and re-reading to be able to better answer your question helped me catch it 👍 ). Unless I missed something, in my testing both json and simplejson will handle unicode properly in `loads` provided they are passed unicode in the first place. So, by moving the decode to the read call ([example](https://github.com/candlepin/python-rhsm/blob/fe0b5f4b2bf730fd9631ddc8a5e81eca0452fef8/src/rhsm/connection.py#L300)), we can knock it out early and avoid traversing the whole object graph.

**Comment by kahowell** [comment]
@awood, re: python 3 as part of the jobs, +1 to that; I'd say let's wait to merge this in until that is done 😄

**Comment by kahowell** [comment]
Rebased on master, and fixed some merge conflicts. @awood, can you please take a quick look at the job added in candlepin/candlepin-jobs@0ea702e (we're just running nosetests with Python 3), and see if that fits the bill? If not, I'm happy to iterate more on it before merging this PR.

**Pull request merged**